### DISCUSSION TITLE
Attaching portmonitor scripts with specified parameters

### DISCUSSION THREAD

**Discussion by pattacini** [robotology/community#105] [Q&A] [answered]
Hi @apaikan 

Do you know whether there's the possibility to specify a parameter in the syntax we use to attach a lua script to a connection?

@Tobias-Fischer is interested in this.

**Comment by apaikan** [comment, marked as answer]
There is not yet a standard way to pass the parameters to a portmonitor plugin (lua/C++) at the connection time due to the limitation imposed by YARP carrier syntax. 

However there are still some ways to do that: 

**Using the YARP administrative commands**
the parameters can be set/configured after establishing the connection via yarp administrative commands.
- See [An example which shows how to use a Lua script to modify image data](http://www.yarp.it/image_modification.html)

**Using the carrier string parameter**
There are also some predefined attributes which can be accessed from the plugin scripts such as 
- `source` : the source port name 
- `destination`: the destination port name 
-  `carrier`: the full carrier string 

These are formatted as the YARP property and given to the portmonitor `create` method: 

``` lua
-- create is called when the port monitor is created 
PortMonitor.create = function(options)
    local sourcePort = options:find("source"):asString()
    print(sourcePort)
    return true;
end
```

Similarly, the `carrier` attribute holds all complete carrier parameter used for connecting ports: 
For example, if a lua plugin is attached using the following command:

``` sh
$ yarp connect /write /read tcp+recv.portmonitor+type.lua+file.codec+myparam.foo
```

Then, the `carrier` attribute holds the  `tcp+recv.portmonitor+type.lua+file.codec+myparam.foo` string. Therefore, a workaround would be to parse this string within the lua code to extract the user-defined parameter (`myparam = foo`): 

``` lua
PortMonitor.create = function(options)
    local carrier = options:find("carrier"):asString()
    print(carrier)
    for attr in string.gmatch(carrier,'([^+]+)') do
        k, v = attr:match("([^.]+).([^.]+)")
        print(k,v)
    end
    return true
end
```

You will get: 

```
t   p
recv    portmonitor
type    lua
file    codec
myparam foo
```

Ignore the `t    p` ;)

**Comment by Tobias-Fischer** [comment]
Thanks @apaikan! That's great. I'll give it a go soon.

**Comment by pattacini** [comment]
Thanks @apaikan for the suggestion!
Closing for now; @Tobias-Fischer feel free to post here again.
